### PR TITLE
support to export gguf q4_0 and q4_1 format

### PR DESCRIPTION
**Pull request by n1ck-guo**
- [x] export function
- [x] q4_0
- [x] q4_1
- [ ] q4_k

using llama.cpp(llama-cli) test q4_0 and q4_1 quantized file, work well.

https://github.com/intel/auto-round/issues/288

### CODE REVIEW

File: auto_round/export/export_to_gguf/quant.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    n = blocks.view(np.uint32)
+    # force nan to quiet
+    n = np.where((n & 0x7fffffff) > 0x7f800000, (n & np.uint32(0xffff0000)) | np.uint32(64 << 16), n)
+    # round to nearest even
+    n = (np.uint64(n) + (0x7fff + ((n >> 16) & 1))) >> 16
+    return n.astype(np.uint16).view(np.uint8)
+
+@register_block("q4_0")
+def q4_0_quant_block(blocks: np.array, scale = None, zp = None):
+    if scale is not None:
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant.py:57]: The symmetry of GGUF differs from what we are currently using., better have a test which one is better

File: auto_round/export/export_to_gguf/export.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    model = kwargs["model"]
+    tokenizer = kwargs.get("tokenizer", None)
+    config = model.config
+
+    tmp_work_dir = Path(os.path.join(output_dir, 'tmp_dir'))
+    if tokenizer is not None:
+        tokenizer.save_pretrained(tmp_work_dir)
+    config.save_pretrained(tmp_work_dir)
+
+    with torch.inference_mode():
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/export.py:49]: have a test of phi-2 and phi-3, , as I recall there were some issues with this context previously.

Comment (n1ck-guo) [reply auto_round/export/export_to_gguf/export.py:49]: Test on phi2, phi-3-mini-4k-instruct. Work well.

File: auto_round/__main__.py [hunk 3]
```diff
@@ -14,9 +14,15 @@
 import sys
 
 def run_eval():
-    from auto_round.script.llm import setup_eval_parser, eval
-    args = setup_eval_parser()
-    eval(args)
+    if "--sequence" in sys.argv:
+        sys.argv.remove("--sequence")
```
Comment (wenhuach21) [inline auto_round/__main__.py]: what's the meaning of this arg

File: auto_round/export/export_to_gguf/convert.py [hunk 4]
```diff
@@ -58,7 +58,7 @@
 
 if 'NO_LOCAL_GGUF' not in os.environ:
     sys.path.insert(1, str(Path(__file__).parent / 'gguf-py'))
-import gguf
+import gguf  # pylint: disable=E0401
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/convert.py]: lazy or hide it in class of function

File: auto_round/script/llm.py [hunk 5]
```diff
@@ -313,8 +313,12 @@ def tune(args):
     from auto_round.utils import logger
 
     if args.format in ["gguf:q4_0", "gguf:q4_1"]:
+        args.bits = 4
+        if args.act_bits <= 8:
+            logger.warning(f"{args.format} not support for activation quantization. Reset act_bits to 16.")
+            args.act_bits = 16
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: remove L319 as there is no such logic in other formats.

File: auto_round/__main__.py [hunk 6]
```diff
@@ -14,15 +14,15 @@
 import sys
 
 def run_eval():
-    if "--sequence" in sys.argv:
-        sys.argv.remove("--sequence")
-        from auto_round.script.llm import setup_eval_parser, eval_sequence
-        args = setup_eval_parser()
-        eval_sequence(args)
-    else:
+    if "--non_sequence" in sys.argv:
```
Comment (wenhuach21) [inline auto_round/__main__.py]: Rename it for easier understanding.

File: auto_round/export/export_to_gguf/export.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        output_type = FTYPE_MAP.get(output_type.lower())
+
+        layer_config = kwargs.get("layer_config")
+
+        model_instance = model_class(model, dir_model=tmp_work_dir, ftype=output_type, fname_out=Path(output_dir),
+                                     layer_config=layer_config, is_big_endian=False, model_name=model_name,
+                                     split_max_tensors=False, split_max_size=0, dry_run=False,
+                                     small_first_shard=False)
+        model_instance.write()
+        logger.info(f"Model successfully exported to {model_instance.fname_out}")
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/export.py:74]: is it possible to serialize tokenizer into model file? gguf models have some issues for tokenizer, right?

File: auto_round/script/llm.py [hunk 8]
```diff
@@ -257,8 +257,8 @@ def tune(args):
     if args.format is None:
         args.format = "auto_round"
     supported_formats = ["auto_round", "auto_gptq", "auto_awq", "auto_round:auto_gptq", "auto_round:auto_awq",
-                         "auto_gptq:marlin", "itrex", "iterx_xpu", "fake"]
-    formats = args.format.replace(' ', '').split(",")
+                         "auto_gptq:marlin", "itrex", "iterx_xpu", "fake", "gguf:q4_0", "gguf:q4_1"]
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: move gguf before itrex

File: auto_round/export/export_to_gguf/export.py (hunk 7 above)
Comment (n1ck-guo) [reply auto_round/export/export_to_gguf/export.py:74]: the Model.set_vocab() is used to serialize tokenizer into model file

File: auto_round/script/llm.py [hunk 9]
```diff
@@ -553,3 +569,60 @@ def eval(args):
...
+
+    if all(s.isdigit() for s in devices):
+        if "CUDA_VISIBLE_DEVICES" in os.environ:
+            current_visible_devices = os.environ["CUDA_VISIBLE_DEVICES"]
+            current_visible_devices = current_visible_devices.split(',')
+            indices = [int(device) for device in devices]
+            try:
+                pick_device = [current_visible_devices[i] for i in indices]
+            except:
+                raise ValueError(
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: reusing the same code later.

File: auto_round/export/export_to_gguf/export.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        "q8_0": gguf.LlamaFileType.MOSTLY_Q8_0,
+        "q4_0": gguf.LlamaFileType.MOSTLY_Q4_0,
+        "q4_1": gguf.LlamaFileType.MOSTLY_Q4_1,
+        "q4_k": gguf.LlamaFileType.MOSTLY_Q4_K_S,
+        "auto": gguf.LlamaFileType.GUESSED,
+    }
+
+def save_quantized_as_gguf(output_dir, backend="gguf:q4_0", **kwargs):
+    """Export the model to gguf format."""
+
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/export.py:39]: add ut and accuracy test

File: test_cuda/test_gguf.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    def test_gguf_format(self):
+        bits, group_size, sym = 4, 32, False
+        autoround = AutoRound(
+            self.model,
+            self.tokenizer,
+            bits=bits,
+            group_size=group_size,
+            sym=sym,
+            iters=2,
+            seqlen=2,
```
Comment (wenhuach21) [inline test_cuda/test_gguf.py:44]: add q4_0

File: auto_round/export/export_to_gguf/convert.py [hunk 12]
```diff
@@ -0,0 +1,4442 @@
...
+
+import math
+import numpy as np
+import torch
+
+if TYPE_CHECKING:
+    from torch import Tensor
+
+if 'NO_LOCAL_GGUF' not in os.environ:
+    sys.path.insert(1, str(Path(__file__).parent / 'gguf-py'))
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/convert.py]: what is this

File: auto_round/export/export_to_gguf/convert.py [hunk 13]
```diff
@@ -0,0 +1,4442 @@
...
+    _model_classes: dict[str, type[Model]] = {}
+
+    dir_model: Path
+    ftype: gguf.LlamaFileType
+    fname_out: Path
+    is_big_endian: bool
+    endianness: gguf.GGUFEndian
+    use_temp_file: bool
+    lazy: bool
+    part_names: list[str]
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/convert.py:88]: why not calling gguf code directly

Comment (n1ck-guo) [reply auto_round/export/export_to_gguf/convert.py:88]: cannot. Different model series use different class to write gguf file.

**Comment by wenhuach21** [comment]
remember to add checker in save_quantized and add warning when combined with fp_layers

**Review by wenhuach21 — APPROVED** [review]

File: auto_round/script/llm.py [hunk 14]
```diff
@@ -253,15 +253,39 @@ def setup_eval_parser():
...
+    from auto_round.utils import detect_device, get_library_version, detect_device_count
+    from auto_round.utils import logger
+
     tasks = args.tasks
     if args.format is None:
         args.format = "auto_round"
     supported_formats = ["auto_round", "auto_gptq", "auto_awq", "auto_round:auto_gptq", "auto_round:auto_awq",
-                         "auto_gptq:marlin", "itrex", "itrex_xpu", "fake"]
-    formats = args.format.replace(' ', '').split(",")
+                         "auto_gptq:marlin", "gguf:q4_0", "gguf:q4_1", "itrex", "iterx_xpu", "fake"]
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: incorrect, you revert the main change

Comment (wenhuach21) [reply auto_round/script/llm.py]: please be very careful when merging

File: auto_round/script/llm.py [hunk 15]
```diff
@@ -253,15 +253,39 @@ def setup_eval_parser():
...
+    formats = args.format.lower().replace(' ', '').split(",")
     for format in formats:
         if format not in supported_formats:
             raise ValueError(f"{format} is not supported, we only support {supported_formats}")
+        if format in ["gguf:q4_0", "gguf:q4_1"]:
```
Comment (wenhuach21) [inline auto_round/script/llm.py:277]: support gguf later if we could inference the exact type by the quantization config

File: auto_round/script/llm.py [hunk 16]
```diff
@@ -253,15 +253,39 @@ def setup_eval_parser():
...
+            args.bits = 4
+        if args.act_bits <= 8:
+            logger.warning(f"{args.format} not support for activation quantization.")
+        if args.group_size != 32:
+            logger.warning(f"{args.format} not support for group_size: {args.group_size}. "
+                "Reset group_size to 32.")
+            args.group_size = 32
+        if args.format.endswith("_0"):
+            args.asym = False
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: add warning if the origin setting is not correct

**Review by wenhuach21 — CHANGES_REQUESTED** [review]
iterx_xpu to itrex_xpu

File: auto_round/autoround.py [hunk 17]
```diff
@@ -1267,6 +1267,14 @@ def save_quantized(self, output_dir=None, format="auto_round", inplace=True, **k
             if processor is not None:
                 processor.save_pretrained(output_dir)
             return
+        if format in ["gguf:q4_0", "gguf:q4_1"]:
+            if self.group_size != 32:
```
Comment (wenhuach21) [inline auto_round/autoround.py:1271]: also better check bits

**Review by wenhuach21 — APPROVED** [review]
